Integrate go-gh API package

samcoe opened the pull request:
This PR replaces our API package implementation with implementation backed by `go-gh` API package. This has resulted in changes across numerous files, because of this I have noted design choices and questions inline with the code.

cc https://github.com/cli/cli/issues/5560

samcoe committed:
Integrate go-gh API package

samcoe commented on api/queries_pr.go:
```diff
@@ -508,34 +503,18 @@ func PullRequestStatus(client *Client, repo ghrepo.Interface, options StatusOpti
-		return "", err
-	}
-
+func pullRequestFragment() string {
```
I updated this function to no longer use feature detection. All the features detected here have exist on GHES 3.0 or earlier. GHES 3.0 was discontinued on 2022-02-16, so we no longer need to support feature detection here and just assume these fields exist in all cases.

samcoe committed:
Fix up tests

samcoe commented on api/client.go:
```diff
@@ -362,110 +132,107 @@ func (c Client) RESTWithNext(hostname string, method string, p string, body io.R
-		return HandleHTTPError(resp)
+// handleResponse takes a ghAPI.HTTPError or ghAPI.GQLError and converts it into an
+// HTTPError or GraphQLError respectively.
+func handleResponse(err error) error {
```
I don't love this name, open to suggestions here.

samcoe commented on cmd/gh/main.go:
```diff
@@ -354,24 +355,22 @@ func checkForUpdate(currentVersion string) (*update.ReleaseInfo, error) {
-		opts = append(opts, api.AddHeader("Authorization", fmt.Sprintf("token %s", token)))
+
+	opts := ghAPI.ClientOptions{Host: ghinstance.Default(), Headers: headers, Log: log}
+	client, err := gh.HTTPClient(&opts)
```
`gh.HTTPClient` provides resolution of the `oauth_token`. Should `basicClient` get moved to the API package?

mislav replied:
I have no qualms with that! But I don't think there will be a significant benefit

samcoe commented on cmd/gh/main.go:
```diff
@@ -354,24 +355,22 @@ func checkForUpdate(currentVersion string) (*update.ReleaseInfo, error) {
-		logTraffic := strings.Contains(debugValue, "api")
-		opts = append(opts, api.VerboseLog(colorable.NewColorable(os.Stderr), logTraffic, colorize))
+	var log io.Writer
+	if debugEnabled, _ := utils.IsDebugEnabled(); debugEnabled {
```
`go-gh` only provides one level of logging so if there is any debug environment variable present then turn on logging.

samcoe commented on internal/authflow/flow.go:
```diff
@@ -141,12 +147,39 @@ func authFlow(oauthHost string, IO *iostreams.IOStreams, notice string, addition
 	return scanner.Err()
 }
+
+func verboseLog(out io.Writer, logTraffic bool, colorize bool) func(http.RoundTripper) http.RoundTripper {
```
This is inlined from `api.VerboseLog` since this was the only reference.

samcoe commented on internal/featuredetection/feature_detection.go:
```diff
@@ -50,9 +48,8 @@ type detector struct {
-	cachedClient := api.NewCachedClient(httpClient, time.Hour*48)
 	return &detector{
-		httpClient: cachedClient,
+		httpClient: httpClient,
```
Detector does not need to be aware of caching, let its callers be responsible for that.

samcoe commented on pkg/cmd/api/api.go:
```diff
@@ -264,13 +266,15 @@ func apiRun(opts *ApiOptions) error {
-	httpClient, err := opts.HttpClient()
+	var httpClient *http.Client
+	if opts.CacheTTL > 0 {
+		httpClient, err = opts.CachedHttpClient(opts.CacheTTL)
```
`CachedClient` can no longer be created from an `HTTPClient`.

samcoe commented on pkg/cmd/api/api_test.go:
```diff
@@ -912,7 +912,7 @@ func Test_apiRun_cache(t *testing.T) {
 	assert.NoError(t, err)
 
-	assert.Equal(t, 1, requestCount)
+	assert.Equal(t, 2, requestCount)
```
Not sure that this test is valid anymore since the caching happens on the `go-gh` side.

samcoe commented on pkg/cmd/factory/default.go:
```diff
@@ -85,7 +86,28 @@ func httpClientFunc(f *cmdutil.Factory, appVersion string) func() (*http.Client,
+	}
+}
+
+func cachedHttpClientFunc(f *cmdutil.Factory, appVersion string) func(time.Duration) (*http.Client, error) {
```
Introduce a way to create cached http clients from the factory.

samcoe commented on pkg/cmd/factory/http.go:
```diff
@@ -2,130 +2,96 @@ package factory
+	SkipAcceptHeaders bool
+}
+
+func NewHTTPClient(opts HTTPClientOptions) (*http.Client, error) {
```
Should this entire file be moved to the API package?

samcoe commented on pkg/cmd/pr/shared/templates.go:
```diff
@@ -129,14 +129,14 @@ type templateManager struct {
 }
 
-func NewTemplateManager(httpClient *http.Client, repo ghrepo.Interface, dir string, allowFS bool, isPR bool) *templateManager {
+func NewTemplateManager(httpClient *http.Client, cachedClient *http.Client, repo ghrepo.Interface, dir string, allowFS bool, isPR bool) *templateManager {
```
This was annoying having to change this function definition to include a cached client, but I couldn't come up with a less intrusive way to make sure the feature detector got a cached client.

mislav commented on api/client_test.go:
```diff
@@ -11,12 +11,21 @@ import (
+	return client
+}
+
+func NewTestClient(reg *httpmock.Registry) *Client {
```
Nit: do `newTest*Client` methods need to be exported? I think they're only ever going to be used in the tests within this package

mislav commented on internal/authflow/flow.go:
```diff
@@ -69,7 +75,7 @@ func authFlow(oauthHost string, IO *iostreams.IOStreams, notice string, addition
 	if debugEnabled {
 		logTraffic := strings.Contains(debugValue, "api")
-		httpClient.Transport = api.VerboseLog(IO.ErrOut, logTraffic, IO.ColorEnabled())(httpClient.Transport)
+		httpClient.Transport = verboseLog(IO.ErrOut, logTraffic, IO.ColorEnabled())(httpClient.Transport)
```
For `oauth.Flow` purposes, a httpClient is anything that responds to `PostForm(string, url.Values) (*http.Response, error)`. Instead of using `http.DefaultClient`, do you think we could wrap a go-gh client (with verbose logging) and wrap it so that it responds to an additional method?

samcoe replied:
I like this idea, but I think it might be a bit out of scope for right now as it would involve making changes to `cli/oauth` repo.

mislav commented on internal/featuredetection/feature_detection_test.go:
```diff
@@ -40,10 +40,11 @@ func TestPullRequestFeatures(t *testing.T) {
-			fakeHTTP := &httpmock.Registry{}
-			httpClient := api.NewHTTPClient(api.ReplaceTripper(fakeHTTP))
+			reg := &httpmock.Registry{}
+			httpClient := http.DefaultClient
```
Let's never mutate `http.DefaultClient` and instead always construct a new `&http.Client{}`

mislav commented on internal/update/update_test.go:
```diff
@@ -71,10 +72,12 @@ func TestCheckForUpdate(t *testing.T) {
-			http := &httpmock.Registry{}
-			client := api.NewClient(api.ReplaceTripper(http))
+			reg := &httpmock.Registry{}
+			httpClient := http.DefaultClient
```
Ditto mutating `http.DefaultClient`

mislav commented on pkg/cmd/factory/http_test.go:
```diff
@@ -43,7 +43,7 @@ func TestNewHTTPClient(t *testing.T) {
 				"authorization": "token MYTOKEN",
 				"user-agent":    "GitHub CLI v1.2.3",
-				"accept":        "application/vnd.github.merge-info-preview+json, application/vnd.github.nebula-preview",
+				"accept":        "application/vnd.github.merge-info-preview+json, application/vnd.github.nebula-preview, application/vnd.github.antiope-preview, application/vnd.github.shadow-cat-preview",
```
If we didn't need `antiope` and `shadow-cat` before, how come they've returned now?

samcoe replied:
`go-gh` adds all the preview headers that could be used regardless of if the host is an enterprise host or not. `antiope` and `shadow-cat` are preview headers for enterprise hosts. As we previously we adding these headers based on the host they also had the same complexities as the authorization token header, in order to reduce complexity I decided it would just be easier to always send them all. From my understanding of the API, this is not harmful and the enterprise ones are just ignored. 

Having said all that, I just double checked the API and neither `antiope`, nor `shadow-cat` exist on GHES 3.1 and above so they probably can just be removed in `go-gh`.

mislav commented on pkg/cmd/issue/delete/delete_test.go:
```diff
@@ -165,7 +165,7 @@ func TestIssueDelete_doesNotExist(t *testing.T) {
 
 	_, err := runCommand(httpRegistry, true, "13")
-	if err == nil || err.Error() != "GraphQL: Could not resolve to an Issue with the number of 13." {
+	if err == nil || err.Error() != "GQL: Could not resolve to an Issue with the number of 13." {
```
Instead of renaming all these error message expectations, how about keeping gh stderr API consistent and rename "GQL" to "GraphQL" in go-gh?

mislav reviewed (changes requested):
Thanks for the gargantuan body of work!

Could there be a way to continue generating a "cached" API client from the go-gh client? That way we wouldn't have to restructure so much code to allow for explicitly passing a cached client around 🤔

samcoe committed:
Address PR comments

samcoe committed:
Remove unused function

samcoe commented:
@mislav I addressed the PR feedback. The tests are going to continue to fail until https://github.com/cli/go-gh/pull/49 gets merged in as that PR has some of the changes you requested. 

The only thing that has not been addressed is the caching client issue. I am open to ideas here as to how to generate a cache client from a normal client, but right now I am running into an order of operations issue. We determine the cache key from the request headers which do not get set right away, they get set somewhere in the transport chain, so adding a caching layer to the top of the transport chain will result in a cache key that is derived from blank headers and won't be unique in some situations, for example if a user refreshes their auth token.

samcoe committed:
Address PR comments

samcoe committed:
Simplify basicClient logic

samcoe committed:
small last cleanup

samcoe committed:
go mod tidy

samcoe commented:
@mislav This is now ready for re-review. After the changes to `go-gh` I was able to remove most of the changes around cached clients which makes this PR a bit less complex.

mislav reviewed (approved):
Fantastic! Thank you for the hard work

samcoe committed:
Catch merge breakage

samcoe committed:
Fix up go mod
